components/renku_data_services/project/api.spec.yaml

```diff
@@ -0,0 +1,382 @@
+          $ref: "#/components/schemas/Visibility"
+        members:
+          $ref: "#/components/schemas/MembersWithRoles"
+      required:
+        - "id"
+        - "name"
+        - "slug"
+        - "created_by"
+
+    Ulid:
+      description: ULID identifier
+      type: string
+      minLength: 26
+      maxLength: 26
+      pattern: "^[A-Za-z0-9]{26}$"
+      format: ulid
+    Name:
+      description: Renku project name
+      type: string
+      minLength: 1
+      maxLength: 99
+      example: "My Renku Project :)"
+    Slug:
+      description: A command-line friendly name for a project
+      type: string
+      minLength: 1
+      maxLength: 99
+      pattern: "^[a-z0-9]+[a-z0-9._-]*$"  # Cannot contain consecutive special characters or end in .git and .atom
+      example: "my-renku-project"
+    CreationDate:
+      description: The date and time the project was created
+      type: string
+      format: date-time
+      example: "2023-11-01T17:32:28Z"
+    User:
+      description: A KeyCloak user
+      type: object
+      additionalProperties: false
+      properties:
+        id:
+          $ref: "#/components/schemas/UserId"
+      required:
+        - "id"
+      example:
+        id: "some-keycloak-user-id"
+    UserId:
+      description: KeyCloak user ID
+      type: string
+      pattern: "^[A-Za-z0-9-]+$"
+      minLength: 1
+      example: "123-keycloak-user-id-456"
+    RepositoriesList:
+      description: A list of repository URLs
+      type: array
+      items:
+        $ref: "#/components/schemas/GitUrl"
+      minItems: 0
+      uniqueItems: true
+    GitUrl:
+      description: URL of a Git repository
+      type: string
+      format: uri  # TODO: This should support other formats like ``git@`` and ``https+git``
+    Visibility:
+      description: Project's visibility levels
+      type: string
+      enum:
+        - "public"
+        - "private"
+    MembersWithRoles:
+      description: List of users and their access level to the project
+      type: array
+      items:
+        $ref: "#/components/schemas/MemberWithRole"
+      minItems: 0
+      uniqueItems: true
+    MemberWithRoles:
+      description: A user and the access level to the project
+      type: object
+      additionalProperties: false
+      properties:
+        user:
+          $ref: "#/components/schemas/User"
+        role:
+          $ref: "#/components/schemas/Role"
+      required:
+        - user
+        - role
+    Role:
+      # TODO: Can have multiple owners? Owner here vs ``created_by``?
+      # TODO: Better names for roles: Maintainer, admin, member, ... instead of reader and writer
+      description: Possible roles of users in a project
+      type: string
+      enum:
+        - "reader"
+        - "writer"
+        - "owner"
+
+    Version:
+      type: object
+      properties:
+        version:
+          type: string
+      required:
+        - "version"
+
+    ErrorResponse:
+      type: object
+      properties:
```

Comment:
What do you think about making this an array field, `errors` which contains one or more `Error` objects, typed as specified here?

Reply:
Sekhar that is an error field that is used throughout the whole API and changing its format will require significant changes to the centralized error handling. Also this error is already returned by other parts of the API and used throughout the UI and other components. I would rather keep this uniform and avoid changing this at this stage.

If we want to change the errors returned we should do it across the whole API all in one go along with the changes needed in the UI and any other components.